contracts/EscrowedNori.sol

```diff
@@ -0,0 +1,725 @@
+ */
+
+uint256 constant SECONDS_IN_TEN_YEARS = 31_536_000;
+
+contract EscrowedNORI is
+  IERC777RecipientUpgradeable,
+  ERC777PresetPausablePermissioned
+{
+  using RemovalUtils for uint256;
+
+  struct EscrowSchedule {
+    uint256 startTime;
+    uint256 endTime;
+    uint256 currentAmount;
+    uint256 claimedAmount;
+    bool exists;
+    uint256 totalQuantityRevoked;
+    uint256 releasedAmountFloor;
+  }
+
+  struct EscrowScheduleDetail {
+    uint256 currentAmount;
+    address recipient;
+    uint256 startTime;
+    uint256 endTime;
+    uint256 claimedAmount;
+    uint256 totalQuantityRevoked;
+    bool exists;
+  }
+
+  /**
+   * @notice Role conferring creation and revocation of escrow schedules.
+   */
+  bytes32 public constant ESCROW_CREATOR_ROLE =
+    keccak256("ESCROW_CREATOR_ROLE");
+
+  /**
+   * @notice Used to register the ERC777TokensRecipient recipient interface in the
+   * ERC-1820 registry
+   *
+   * @dev Registering that EscrowedNORI implements the ERC777TokensRecipient interface with the registry is a
+   * requirement to be able to receive ERC-777 BridgedPolygonNORI tokens. Once registered, sending BridgedPolygonNORI
+   * tokens to this contract will trigger tokensReceived as part of the lifecycle of the BridgedPolygonNORI transaction
+   */
+  bytes32 public constant ERC777_TOKENS_RECIPIENT_HASH =
+    keccak256("ERC777TokensRecipient");
+
+  mapping(address => uint256[]) private _addressToEscrowScheduleIds;
+
+  mapping(address => mapping(uint256 => EscrowSchedule))
+    private _addressToEscrowSchedules;
+
+  /**
+   * @notice The BridgedPolygonNORI contract that this contract wraps tokens for
+   */
+  BridgedPolygonNORI private _bridgedPolygonNori;
+
+  /**
+   * @notice The Removal contract that accounts for supply.
+   */
+  Removal private _removal;
+
+  /**
+   * @notice The [ERC-1820](https://eips.ethereum.org/EIPS/eip-1820) pseudo-introspection registry
+   * contract
+   *
+   * @dev Registering that EscrowedNORI implements the ERC777TokensRecipient interface with the registry is a
+   * requiremnt to be able to receive ERC-777 BridgedPolygonNORI tokens. Once registered, sending BridgedPolygonNORI
+   * tokens to this contract will trigger tokensReceived as part of the lifecycle of the BridgedPolygonNORI transaction
+   */
+  IERC1820RegistryUpgradeable private _erc1820;
+
+  /**
+   * @notice Emitted on successful creation of a new escrow schedule.
+   */
+  event EscrowScheduleCreated(
+    address indexed recipient,
+    uint256 indexed startTime
+  );
+
+  /**
+   * @notice Emitted on when unreleased tokens of an active escrow schedule are revoked.
+   */
+  event UnreleasedTokensRevoked(
+    uint256 indexed atTime,
+    uint256 indexed removalId,
+    uint256 indexed escrowScheduleId,
+    uint256 quantity
+  );
+
+  /**
+   * @notice Emitted on withdwal of fully unlocked tokens.
+   */
+  event TokensClaimed(
+    address indexed from,
+    address indexed to,
+    uint256 quantity
```

Comment:
What's the reason for not indexing quantity?

Reply:
oh we actually can - this is how the event looks in `LockedNORI` so I just didn't modify it. We'll almost certainly never use that index since quantity can be duplicate across different withdrawals and therefore isn't a very interesting filter, but I suppose it doesn't hurt to index since we can have 3